File: hola-mundo-backend-2022-grupo-7/src/main/kotlin/trabajoPracticoAlgoritmos2/Destino.kt
package trabajoPracticoAlgoritmos2

import funcionesAuxiliaresYConstantes.coincidenciaParcial

class Destino(
    var pais: String,
    var ciudad: String,
    var costo: Double,
    override var id: Int = -1,
    ): ElementosDelRepositorio {

    fun esLocal(): Boolean {
        return pais.uppercase() == PAIS_LOCAL
    }

    fun calcularCostoBase(): Double {

        if (!esLocal()) {
            return costo + costo * BONUS_POR_NO_SER_DESTINO_LOCAL
        } else {
            return costo
        }
    }

    override fun condicionDeBusqueda(textoDeBusqueda: String): Boolean {
        return this.busquedaPorCiudad(textoDeBusqueda) || this.busquedaPorPais(textoDeBusqueda)
    }

    fun busquedaPorCiudad(textoDeBusqueda: String):Boolean{         //estos metodos estan solo para facilitar el testeo de cada una de las
        return coincidenciaParcial(this.ciudad,textoDeBusqueda)        // clases de equivalencia
    }

    fun busquedaPorPais(textoDeBusqueda: String):Boolean{
        return coincidenciaParcial(this.pais,textoDeBusqueda)
    }

    override fun validacion(): Boolean {

        if (pais.isNullOrBlank() || ciudad.isNullOrBlank() || costo <= 0) {
            return false
        }
        return true
    }

}

File: tpreact/src/Domain/Destino.js
import {PAIS_LOCAL, BONUS_POR_NO_SER_DESTINO_LOCAL} from "./Constantes"

export class DestinoJson {

    id
    pais = ""
    ciudad = ""
    costo
}

export class Destino{
    
    id
    pais 
    ciudad 
    costo

    constructor(_pais = "", _ciudad = "", _costo = 0){
        this.pais = _pais
        this.ciudad = _ciudad
        this.costo = _costo
    }

    static fromJson(destinoJson){
        return Object.assign(new Destino(),destinoJson)
    }

    esLocal(){
        return this.pais.toUpperCase() == PAIS_LOCAL
    }

    calcularCostoBase(){

        let costoAux = this.costo

        if (!this.esLocal()) {
            return costoAux + (costoAux * BONUS_POR_NO_SER_DESTINO_LOCAL)
        }
        return costoAux
    }

    validacion() {     //Si el pais o la ciudad estan vacios o son null, esto devuelve false
        return !(!this.pais || !this.ciudad || this.costo <= 0)
    }

}